picker_wrapper.go

```diff
@@ -129,6 +135,18 @@ func (pw *pickerWrapper) pick(ctx context.Context, failfast bool, info balancer.
 			continue
 		}
 
+		// If the channel is set, it means that the pick call had to wait for a
+		// new picker at some point. Either there was no picker and it received
+		// a new one, or a picker errored with ErrNoSubConnAvailable or errored
+		// with failfast set to false, which will trigger a continue to the next
+		// iteration. In that second case the only way it gets to this codeblock
+		// is to receive a new picker either in UpdateState or the select above.
```

Comment:
This is confusing because a new picker is _always_ received by `UpdateState`, not by the `select`.  Maybe this sentence is: "In the latter case, we only reach this when there is a new picker."

And isn't this true in either of the two cases?  Basically, we only reach this point when: 1. there is a picker immediately on the first iteration, or 2. we have a new picker available.  That's why we grab the picker on the following lines.  The `if ch != nil` filters out the first case where we never needed to block.

Reply:
Yeah, I guess the language I had is technically incorrect, but I was going for it either UpdateState() before the next iteration or it UpdateState() while blocking on the <-ch if the new one didn't come before the for. Yeah, the first case (either it's there, or it's not and we block) gets there, with it being there already not triggering conditional, and not being there triggering conditional. However, I think that it is more easily implied by conditional, but let me reword the comment and see if you like it more technically.